clang/lib/CIR/CodeGen/CIRGenExprScalar.cpp

```diff
@@ -1012,6 +1013,88 @@ static std::optional<QualType> getUnwidenedIntegerType(const ASTContext &Ctx,
          (2 * Ctx.getTypeSize(RHSTy)) < PromotedSize;
 }
 
+/// Emit pointer + index arithmetic.
+static mlir::Value buildPointerArithmetic(CIRGenFunction &CGF,
+                                          const BinOpInfo &op,
+                                          bool isSubtraction) {
+  // Must have binary (not unary) expr here.  Unary pointer
+  // increment/decrement doesn't use this path.
+  const BinaryOperator *expr = cast<BinaryOperator>(op.E);
+
+  mlir::Value pointer = op.LHS;
+  Expr *pointerOperand = expr->getLHS();
+  mlir::Value index = op.RHS;
+  Expr *indexOperand = expr->getRHS();
+
+  // In a subtraction, the LHS is always the pointer.
+  if (!isSubtraction && !pointer.getType().isa<mlir::cir::PointerType>()) {
+    std::swap(pointer, index);
+    std::swap(pointerOperand, indexOperand);
+  }
+
+  bool isSigned = indexOperand->getType()->isSignedIntegerOrEnumerationType();
+
+  auto &DL = CGF.CGM.getDataLayout();
+
+  // Some versions of glibc and gcc use idioms (particularly in their malloc
+  // routines) that add a pointer-sized integer (known to be a pointer value)
+  // to a null pointer in order to cast the value back to an integer or as
+  // part of a pointer alignment algorithm.  This is undefined behavior, but
+  // we'd like to be able to compile programs that use it.
+  //
+  // Normally, we'd generate a GEP with a null-pointer base here in response
+  // to that code, but it's also UB to dereference a pointer created that
+  // way.  Instead (as an acknowledged hack to tolerate the idiom) we will
+  // generate a direct cast of the integer value to a pointer.
+  //
+  // The idiom (p = nullptr + N) is not met if any of the following are true:
+  //
+  //   The operation is subtraction.
+  //   The index is not pointer-sized.
+  //   The pointer type is not byte-sized.
+  //
+  if (BinaryOperator::isNullPointerArithmeticExtension(
+          CGF.getContext(), op.Opcode, expr->getLHS(), expr->getRHS()))
+    llvm_unreachable("null pointer arithmetic extension is NYI");
+
+  if (UnimplementedFeature::dataLayoutGetIndexTypeSizeInBits()) {
+    // TODO(cir): original codegen zero/sign-extends the index to the same width
+    // as the pointer. Since CIR's pointer stride doesn't care about that, it's
+    // skiped here.
+    llvm_unreachable("target-specific pointer width is NYI");
+  }
+
+  // If this is subtraction, negate the index.
+  if (isSubtraction)
+    index = CGF.getBuilder().createNeg(index);
```

Comment:
LLVM doesn't differentiate signed vs unsigned, for now in CIR it'd be good to add an assert here that the index actually got mapped to a signed `mlir::cir::IntType`.

Reply:
Nice catch!

As a matter of fact, I think the `createNeg` should ensure a signed integer as the resulting type. I'll make sure it works for integers and I'll also add a `llvm_unreacheable` for NYI types.